Merge multi-precision support into main (#213)
BerkeleyLab/inference-engine: README.md

Because this repository supports programming language research, the code exercises new language features in novel ways.
We recommend using any compiler's latest release or even building open-source compilers from source.
The [handy-dandy] repository contains scripts capturing steps for building the [LLVM] compiler suite.
The remainder of this section contains commands for building Fiats with a recent Fortran compiler and the Fortran Package Manager ([`fpm`]).

#### Supported Compilers
##### LLVM (`flang-new`)
With LLVM `flang` 20 installed in your `PATH`, build and test Fiats with the installed `flang-new` symlink in order for `fpm` to correctly identify the compiler:
```
fpm test --compiler flang-new --flag "-O3"
```
With LLVM `flang` 19, enable the compiler's experimental support for assumed-rank entities:

```
fpm test --compiler flang-new --flag "-mmlir -allow-assumed-rank -O3"
```

###### _Experimental:_ Automatic parallelization of `do concurrent` on CPUs

> With the `amd_trunk_dev` branch of the [ROCm fork] fork of LLVM, automatically parallelize inference calculations inside `do concurrent` constructs:

Reviewer: ```suggestion
With the `amd-trunk-dev` branch of the [ROCm fork] of LLVM, automatically parallelize inference calculations inside `do concurrent` constructs:
```

  -- --network model.json

```
where `model.json` must be a neural network in the [JSON] format used by Fiats and the companion [nexport] package.

Automatic parallelization for training neural networks is under development.

#### Partially Supported Compilers

Fiats release 0.14.0 and earlier support the use of the NAG, GNU, and Intel Fortran compilers.
We are corresponding with these compilers' developers about addressing the compiler issues preventing building newer Fiats releases.

##### NAG (`nagfor`)
```
fpm test --compiler nagfor --flag -fpp --profile release
```

##### GNU (`gfortran`)
Compiler bugs related to parameterized derived types currently prevent `gfortran` from building Fiats versions 0.15.0 or later.
Test and build earlier versions of Fiats build with the following command:
```
fpm test --compiler gfortran --profile release
```

##### Intel (`ifx`)

> Compiler bugs related to parameterized derived types currently prevent `gfortran` from building Fiats versions 0.15.0 or later.

Reviewer: This line reference gfortran but is talking about ifx. Is the below suggestion sufficient to be accurate to the current situation?

```suggestion
Compiler bugs related to parameterized derived types currently prevent `ifx` from building Fiats versions 0.15.0 or later.
```

Author: ```suggestion
Compiler bugs related to generic name resolution currently prevent `ifx` from building Fiats versions 0.15.0 or later.
```

Test and build earlier versions of Fiats build with the following command:
```
fpm test --compiler ifx --profile release --flag -O3
```

##### _Experimental:_ Automatic offloading of `do concurrent` to GPUs
This capability is under development with the goal to facilitate automatic GPU offloading via the following command:
```
fpm test --compiler ifx --profile release --flag "-fopenmp-target-do-concurrent -qopenmp -fopenmp-targets=spir64 -O3"
```

#### Under Development

> We are corresponding with the developers of the compiler(s) below developers about addressing the compiler issues preventing building newer Fiats releases.

Reviewer: ```suggestion
We are corresponding with the developers of the compiler(s) below about addressing the compiler issues preventing building Fiats.
```
